### PR TITLE
fix(aws-cdk-lib): allow create explicitly empty trees on aws-cdk-lib core

### PR DESCRIPTION
**Pull request by Rondineli**
### Issue # (if applicable)

Closes #[29165](https://github.com/aws/aws-cdk/issues/29165).

### Reason for this change

Using `addOverride` calls `deepMerge`, and few edge cases like wafv2 rules needs Actions and other keys with objects and empty objects.

### Description of changes

Changing the code to be more realistic with the comment, where only null data is deleted

### Description of how you validated changes

No tests, just making sure the current change would be compatible with current tests, should not break anything

### Checklist
- [ x ] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by jeffersonfrazer** [comment]
+1

File: packages/aws-cdk-lib/core/lib/cfn-resource.ts [hunk 1]
```diff
@@ -654,7 +664,6 @@ function deepMerge(target: any, ...sources: any[]) {
       if (key === '__proto__' || key === 'constructor') {
         continue;
       }
-
```
Comment (paulhcsun) [inline packages/aws-cdk-lib/core/lib/cfn-resource.ts:657]: can you add back the linebreak here

File: packages/aws-cdk-lib/core/lib/cfn-resource.ts [hunk 2]
```diff
@@ -640,6 +640,16 @@ const MERGE_EXCLUDE_KEYS: string[] = [
   'Fn::Or',
 ];
 
+// Availabe wafv2 empty objects
+// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-wafv2-webacl-ruleaction.html
```
Comment (paulhcsun) [inline packages/aws-cdk-lib/core/lib/cfn-resource.ts:644]:
```suggestion
/**
 * Availabe WAFv2 empty objects
 *
 * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-wafv2-webacl-ruleaction.html
 */
```

File: packages/aws-cdk-lib/core/lib/cfn-resource.ts [hunk 3]
```diff
@@ -640,6 +640,16 @@ const MERGE_EXCLUDE_KEYS: string[] = [
   'Fn::Or',
 ];
 
+// Availabe wafv2 empty objects
```
Comment (paulhcsun) [inline packages/aws-cdk-lib/core/lib/cfn-resource.ts:643]: This description is vague and doesn't explain what these are used for, nor how "empty objects" relates to `DELETE_MERGE_EXCLUDE_KEYS`. Maybe provide a short snippet of context, you can reference the description for `MERGE_EXCLUDE_KEYS` on line 617

File: packages/aws-cdk-lib/core/test/resource.test.ts [hunk 4]
```diff
@@ -541,6 +541,44 @@ describe('resource', () => {
       });
     });
 
+    test('addOverride(p, {object: {object: {}}}) will respect empty trees', () => {
```
Comment (paulhcsun) [inline packages/aws-cdk-lib/core/test/resource.test.ts:544]: prefer different wording like `support` over `respect` here

File: packages/aws-cdk-lib/core/lib/cfn-resource.ts [hunk 5]
```diff
@@ -725,7 +734,20 @@ function deepMerge(target: any, ...sources: any[]) {
         // sibling concrete values alongside, so we can delete this tree.
         const output = target[key];
         if (typeof(output) === 'object' && Object.keys(output).length === 0) {
-          delete target[key];
+          /*
+          * some empty trees are needed in some edge cases
```
Comment (paulhcsun) [inline packages/aws-cdk-lib/core/lib/cfn-resource.ts:738]:
```suggestion
          * Some empty trees are needed in some edge cases
```

File: packages/aws-cdk-lib/core/lib/cfn-resource.ts [hunk 6]
```diff
@@ -725,7 +734,20 @@ function deepMerge(target: any, ...sources: any[]) {
...
+          * ie: wafv2 Rules (for Action, OverrideAction, etc...)
```
Comment (paulhcsun) [inline packages/aws-cdk-lib/core/lib/cfn-resource.ts:739]: Please clarify which `some empty trees` and maybe combine the `ie` into the `some edge cases` like `... are needed when WAFv2 rule action is one of 'Action', 'Block', 'Captcha', 'Challenge' or 'Count'.`

**Review by paulhcsun — CHANGES_REQUESTED** [review]
Left some comments

**Comment by paulhcsun** [comment]
@Rondineli I have some concerns about this change breaking existing customers so please add it behind a [feature flag](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md?rgh-link-date=2024-03-12T16%3A45%3A17Z#feature-flags).

**Comment by Rondineli** [comment]
> @Rondineli I have some concerns about this change breaking existing customers even behind a feature flag. Please have a look at the comments on the original issue.

Sorry, but I really dont get the backwards compatibility bit? As I am explicitly setting the keys to it ignore. And trust me, I tried the way it is right now and cloudformation broke because those missing keys on waf rules, so, my guess is if someone tried to use it, it did not work.

I would be happy to change the PR for another approach, but this is definitely something broken and wont cause regression, quite the opposite.

**Pull request closed without merging**